src/string/grapheme/properties.cr

```diff
@@ -0,0 +1,1670 @@
+module String::Grapheme
+  # The unicode properties. Only the ones needed in the context of this shard are included.
+  private enum Property
+    Any
+    Preprend
+    CR
+    LF
+    Control
+    Extend
+    RegionalIndicator
+    SpacingMark
+    L
+    V
+    T
+    LV
+    LVT
+    ZWJ
+    ExtendedPictographic
+
+    # returns the Unicode property value (see `Codepoints` constants below)
+    # of the given code point
+    def self.from(r : Int32)
+      # run a binary search
+      f = 0
+      t = Grapheme.codepoints.size
+      while t > f
+        mid = (f + t) // 2
+        cp = Grapheme.codepoints[mid]
+        if r < cp[0]
+          t = mid
+          next
+        end
+        if r > cp[1]
+          f = mid + 1
+          next
+        end
+        return cp[2]
+      end
+      Property::Any
+    end
+  end
+
+  # Maps code point ranges to their properties. In the context of this package,
+  # any code point that is not contained may map to "prAny". The code point
+  # ranges in this slice are numerically sorted.
+  #
+  # These ranges were taken from
+  # http://www.unicode.org/Public/UCD/latest/ucd/auxiliary/GraphemeBreakProperty.txt
+  # as well as
+  # https://unicode.org/Public/emoji/latest/emoji-data.txt
+  # ("Extended_Pictographic" only) on March 11, 2019. See
+  # https://www.unicode.org/license.html for the Unicode license agreement.
+  @@codepoints : Array(Tuple(Int32, Int32, Property))?
+
+  # :nodoc:
+  protected def self.codepoints
+    @@codepoints ||= begin
+      data = Array(Tuple(Int32, Int32, Property)).new
+
+      put(data, {0x0000, 0x0009, Property::Control})                # Cc  [10] <control-0000>..<control-0009>
```

Comment:
Would it be possible to autogenerate this list from the actual file as provided by unicode.org, perhaps using a macro?  My concern is that this will be difficult to keep up to date as new emoji versions come out, but if we can automate it, then generating it is as easy as downloading the latest text file.